network-config-analyzer/GenericYamlParser.py

```diff
@@ -46,30 +46,45 @@ def warning(self, msg, obj=None):
         For keys that get specific known values only, it checks that the existing value of the key in the given dict
         is in the allowed_values.
         Allowed values contains all the possible values of a key in the correct type (value and type check).
         :param dict dict_to_check: The dictionary for which the keys should be checked
         :param str dict_name: A name for the dictionary (providing context in error messages)
-        :param dict allowed_keys: Map from allowed keys to usage code
-        (0-optional, 1-must have, 2-not yet supported)
+        :param dict allowed_keys: Map from allowed keys to usage code (0-optional, 1-must have, 2-not yet supported)
+         and type (type is optional) - if type is specified, allowed_keys maps string to array.
         :param dict allowed_values: Map from a key name to its allowed values (optional)
         :return: None
         :raises SyntaxError: if some of the keys are not allowed/missing
         """
-        if dict_to_check:
-            if not isinstance(dict_to_check, dict):
-                self.syntax_error('type of ' + dict_name + 'is not a map', dict_to_check)
-        for key, code in allowed_keys.items():
+
+        for key, key_info in allowed_keys.items():
+            if isinstance(key_info, list):
+                code = key_info[0]
+                value_type = key_info[1]
+            else:
+                code = key_info
+                value_type = None
             if code == 1 and key not in dict_to_check:
-                self.syntax_error(f'{dict_name} must have a {key} entry', dict_to_check)
-            if code == 2 and key in dict_to_check:
-                self.syntax_error(f'{key} is not yet supported inside {dict_name}', dict_to_check)
-            if key in dict_to_check and key in allowed_values:
+                self.syntax_error(f'{dict_name} must have {key} entry', dict_to_check)
+            if key in dict_to_check:
+                if code == 2:
+                    self.syntax_error(f'{key} is not yet supported inside {dict_name}', dict_to_check)
                 value = dict_to_check.get(key)
-                if value not in allowed_values[key]:
-                    self.syntax_error(f'{key} has invalid value/type in {dict_name}')
+                if code == 1 and value is None:
+                    self.syntax_error(f'mandatory {key} entry can not be null in {dict_name}', dict_to_check)
+                if value_type is not None:
+                    if value == 0 and not value_type == int:
+                        self.syntax_error(f'type of {key} is not {value_type} in {dict_name}',
+                                          dict_to_check)
+                    if value is not None and not isinstance(value, value_type):
+                        self.syntax_error(f'type of {key} is not {value_type} in {dict_name}',
+                                          dict_to_check)
+                if key in allowed_values:
```

Comment:
I'm not sure, but isn't this should be `if key in dict_to_check and key in allowed_values:`

Reply:
as mentioned above, this if statement is included under if key in dict_to_check